File: apps/ios/WindmillKit/Sources/WindmillJournal/PageCache.swift
import Foundation
import WindmillPlatform

// One file per seat, the seat in the file name, so no seat's store can read another seat's bytes.
// Every write lands here marked `needsPush` and loses the mark only once the server answers.

public final class PageCache {
    public struct Entry: Codable, Equatable {
        public var page: Page
        public var needsPush: Bool
    }

    private let url: URL
    private var entries: [String: Entry]

    // Read once on open and retired, never written again.
    private static let legacyFile = "windmill-journal-pages.json"
    // Where its unsent pages wait; no seat opens this name.
    private static let quarantineFile = "windmill-journal-pages-unclaimed.json"

    public init(url: URL) {
        self.url = url
        let data = (try? Data(contentsOf: url)) ?? Data()
        entries = (try? JSONDecoder().decode([String: Entry].self, from: data)) ?? [:]
    }

    // `seat` is the signed-in user's id, nil for nobody signed in. `deviceHolds` is who this device
    // holds a SESSION for — the Keychain's answer, not this seat's.
    public convenience init(seat: String?, in directory: URL = PageCache.deviceDirectory(),
                            deviceHolds: @escaping @autoclosure () -> String? = KeychainSessions().readUser()?.id) {
        PageCache.retireLegacyStore(in: directory, heldBy: deviceHolds)
        if let seat { PageCache.carryOverTheUnprefixedFile(for: seat, in: directory) }
        self.init(url: directory.appendingPathComponent(PageCache.fileName(forSeat: seat)))
    }

    public static func deviceDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    // The seat id is a path component: anything not plainly an id is dropped, or a separator or `..`
    // opens another seat's file. The `u.` prefix keeps a seat named `anon` or `unclaimed` off those files.
    static func fileName(forSeat seat: String?) -> String {
        guard let seat else { return "windmill-journal-pages-anon.json" }
        let safe = seat.filter { $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "-" || $0 == "_") }
        return "windmill-journal-pages-u.\(safe).json"
    }

    // Delete once no build writing the un-prefixed name is in anybody's hands.
    private static func carryOverTheUnprefixedFile(for seat: String, in directory: URL) {
        let safe = seat.filter { $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "-" || $0 == "_") }
        let unprefixed = directory.appendingPathComponent("windmill-journal-pages-\(safe).json")
        guard FileManager.default.fileExists(atPath: unprefixed.path) else { return }
        let theirs = PageCache(url: directory.appendingPathComponent(fileName(forSeat: seat)))
        for (_, entry) in decode(unprefixed) { theirs.store(entry.page, needsPush: entry.needsPush) }
        guard theirs.flush() else { return }
        try? FileManager.default.removeItem(at: unprefixed)
    }

    // Unsent pages go to the seat whose SESSION this device holds, or to quarantine if it holds none.
    // The file is removed only once they have landed, so a refused write is retried on the next open.
    static func retireLegacyStore(in directory: URL, heldBy deviceHolds: () -> String?) {
        let legacy = directory.appendingPathComponent(legacyFile)
        guard let data = try? Data(contentsOf: legacy) else { return }
        let held = (try? JSONDecoder().decode([String: Entry].self, from: data)) ?? [:]
        let owed = held.filter { $0.value.needsPush }

        guard keep(owed, in: directory, heldBy: deviceHolds()) else { return }
        try? FileManager.default.removeItem(at: legacy)
    }

    private static func keep(_ owed: [String: Entry], in directory: URL, heldBy seat: String?) -> Bool {
        guard let seat else {
            let quarantine = directory.appendingPathComponent(quarantineFile)
            var waiting = decode(quarantine)
            for (day, entry) in owed where waiting[day] == nil { waiting[day] = entry }
            guard let encoded = try? JSONEncoder().encode(waiting) else { return false }
            do {
                try encoded.write(to: quarantine, options: .atomic)
            } catch {
                return false
            }
            return true
        }
        let theirs = PageCache(url: directory.appendingPathComponent(fileName(forSeat: seat)))
        for entry in owed.values { theirs.store(entry.page, needsPush: true) }
        return theirs.flush()
    }

    // Reading is not claiming: nothing releases these to a seat without a person saying so.
    public static func quarantinedPages(in directory: URL = PageCache.deviceDirectory()) -> [Page] {
        decode(directory.appendingPathComponent(quarantineFile))
            .values.map(\.page).sorted { $0.day < $1.day }
    }

    private static func decode(_ url: URL) -> [String: Entry] {
        let data = (try? Data(contentsOf: url)) ?? Data()
        return (try? JSONDecoder().decode([String: Entry].self, from: data)) ?? [:]
    }

    public var pages: [Page] {
        entries.values.map(\.page).sorted { $0.day < $1.day }
    }

    public func page(on day: LocalDay) -> Page? {
        entries[day.iso]?.page
    }

    // Oldest first, so a claim replays in order.
    public var pending: [Page] {
        entries.values.filter(\.needsPush).map(\.page).sorted { $0.day < $1.day }
    }

    // Keeps whichever stamp is greater, as the server decides it.
    @discardableResult
    public func store(_ page: Page, needsPush: Bool) -> Page {
        let held = entries[page.day.iso]
        let winner = Page.winner(of: page, and: held?.page)

        // A local edit always owes; a remote page settles the debt only by beating what was held.
        let heldPageSurvived = held?.needsPush == true && winner.stamp == held?.page.stamp
        entries[page.day.iso] = Entry(page: winner, needsPush: needsPush || heldPageSurvived)
        return winner
    }

    // An ack must not clear the debt outright: a newer local write may be waiting and is still owed.
    public func markPushed(_ day: LocalDay, winner: Page) {
        let held = entries[day.iso]
        // Strictly newer, not "at least as new": an ack usually returns the stamp just sent.
        let newerWriteIsWaiting = held?.needsPush == true && (held?.page.stamp ?? .zero) > winner.stamp
        entries[day.iso] = Entry(page: Page.winner(of: winner, and: held?.page), needsPush: newerWriteIsWaiting)
    }

    // Answers whether the device took the bytes.
    @discardableResult
    public func flush() -> Bool {
        guard let data = try? JSONEncoder().encode(entries) else { return false }
        do {
            try data.write(to: url, options: .atomic)
            return true
        } catch {
            return false
        }
    }

    // Call only once the arriving seat's file has taken the drafts and flushed them.
    public func discard() {
        entries = [:]
        try? FileManager.default.removeItem(at: url)
    }
}

File: web/src/products/journal/pageCache.js
// One localStorage key per scope, read whole at construction and written whole on flush; an unreadable
// store opens empty. Two marks per entry: `needsPush` — this device owes the account this page; `read` —
// this device has seen the account's answer for that day. Owed and not read carries no stamp, because a
// stamp claims to have written last: `store` takes stamped pages, `hold` the unstamped one.
// The scope is the key — `wm.journal.pages.u.<userId>`, or `wm.journal.pages.anon` signed out — and a
// cache opened for one account never reads another account's bytes.

import { compareStamps, ZERO_STAMP } from './hlc.js';

// The unscoped store: read once on open, retired into quarantine, never written again.
const LEGACY_KEY = 'wm.journal.pages';
const ANON_KEY = 'wm.journal.pages.anon';
// No scope opens this key: nothing here is drawn, indexed or sent until a signed-in person restores it.
const UNCLAIMED_KEY = 'wm.journal.pages.unclaimed';

export function keyForScope(account) {
  return account ? `wm.journal.pages.u.${account}` : ANON_KEY;
}

// A flush keeps everything owed whatever its age, plus the newest RETAIN_DAYS entries.
const RETAIN_DAYS = 120;

// Private-mode browsers throw on first use and Node has no localStorage: both read as no device tier.
export function deviceStorage(key = ANON_KEY) {
  try {
    const storage = globalThis.localStorage;
    storage.getItem(key);
    return storage;
  } catch {
    return null;
  }
}

// A day the writer touched at all: a mood with no words counts.
export function isWritten(page) {
  return Boolean(page.body) || page.mood != null || page.energy != null;
}

export function blankPage(day) {
  return { day, body: '', mood: null, energy: null, source: 'typed', stamp: ZERO_STAMP };
}

// On the wire an unset scale is 0; inside this module an unset scale is null and nothing else.
export function normalizePage(raw) {
  return {
    day: String(raw?.day ?? ''),
    body: typeof raw?.body === 'string' ? raw.body : '',
    mood: raw?.mood || null,
    energy: raw?.energy || null,
    source: raw?.source === 'spoken' ? 'spoken' : 'typed',
    stamp: typeof raw?.stamp === 'string' ? raw.stamp : ZERO_STAMP,
  };
}

// Last writer wins by stamp; ties go to the page already held.
export function winnerOf(incoming, held) {
  if (!held) return incoming;
  return compareStamps(incoming.stamp, held.stamp) > 0 ? incoming : held;
}

function readEntries(storage, key) {
  if (!storage) return [];
  try {
    const stored = JSON.parse(storage.getItem(key));
    if (!stored || typeof stored !== 'object') return [];
    return Object.entries(stored)
      .filter(([day, entry]) => day && entry && entry.page)
      .map(([day, entry]) => [day, {
        page: { ...normalizePage(entry.page), day },
        needsPush: Boolean(entry.needsPush),
        read: Boolean(entry.read),
      }]);
  } catch {
    return [];
  }
}

// Owed entries go to quarantine: "unsent" says nothing about who wrote them. Everything else is an
// unattributable cached copy and is dropped. A storage that refuses the write keeps the legacy key.
function retireLegacyStore(storage) {
  try {
    if (storage.getItem(LEGACY_KEY) === null) return;
    const quarantined = {};
    for (const [day, entry] of readEntries(storage, UNCLAIMED_KEY)) quarantined[day] = entry;
    for (const [day, entry] of readEntries(storage, LEGACY_KEY)) {
      if (!entry.needsPush || quarantined[day]) continue;
      quarantined[day] = { page: { ...entry.page, stamp: ZERO_STAMP }, needsPush: true, read: false };
    }
    storage.setItem(UNCLAIMED_KEY, JSON.stringify(quarantined));
    storage.removeItem(LEGACY_KEY);
  } catch {
    // Refused: the legacy key stays as it is and the next open retries.
  }
}

// What is waiting in quarantine, oldest first. Reading is not claiming.
export function unclaimedPages(storage = deviceStorage()) {
  if (!storage) return [];
  return readEntries(storage, UNCLAIMED_KEY)
    .map(([, entry]) => ({ ...entry.page, stamp: ZERO_STAMP }))
    .sort((left, right) => (left.day < right.day ? -1 : 1));
}

// Emptied on a restore that landed, and on a discard.
export function dropUnclaimedPages(storage = deviceStorage()) {
  try { storage.removeItem(UNCLAIMED_KEY); } catch { /* no device tier */ }
}

// Only owed drafts travel, and unstamped: nothing written signed out has read an account's page.
export function anonymousDrafts(storage = deviceStorage()) {
  if (!storage) return [];
  return readEntries(storage, ANON_KEY)
    .filter(([, entry]) => entry.needsPush)
    .map(([, entry]) => ({ ...entry.page, stamp: ZERO_STAMP }))
    .sort((a, b) => (a.day < b.day ? -1 : 1));
}

// Called only once the account's scope has taken the drafts and flushed them.
export function dropAnonymousScope(storage = deviceStorage()) {
  try { storage.removeItem(ANON_KEY); } catch { /* no device tier */ }
}

export class PageCache {
  constructor(account, storage = deviceStorage()) {
    this.storage = storage;
    this.key = keyForScope(account ?? null);
    if (storage) retireLegacyStore(storage);
    this.entries = new Map(readEntries(storage, this.key));
  }

  page(day) {
    return this.entries.get(day)?.page ?? null;
  }

  // Whether this device has seen the account's answer for a day; a stamp is minted only where true.
  hasRead(day) {
    return this.entries.get(day)?.read === true;
  }

  pages() {
    return [...this.entries.values()].map((entry) => entry.page).sort((a, b) => (a.day < b.day ? -1 : 1));
  }

  // Everything owed, oldest first. Entries, not pages: the sender must know whether the day was read.
  owed() {
    return [...this.entries.values()].filter((entry) => entry.needsPush).sort((a, b) => (a.page.day < b.page.day ? -1 : 1));
  }

  // A stamped page arriving from anywhere; convergence is decided as the server decides it.
  store(page, { needsPush = false, read = false } = {}) {
    const held = this.entries.get(page.day) ?? null;
    const winner = winnerOf(page, held?.page ?? null);
    // An arriving page settles the debt only by beating what was held.
    const heldSurvived = Boolean(held?.needsPush) && winner === held.page;
    this.entries.set(page.day, {
      page: winner,
      needsPush: needsPush || heldSurvived,
      read: read || Boolean(held?.read),
    });
    return winner;
  }

  // A day this device has never read: no stamp, no race — it replaces what was held.
  hold(draft) {
    this.entries.set(draft.day, {
      page: { ...normalizePage(draft), stamp: ZERO_STAMP },
      needsPush: true,
      read: false,
    });
  }

  // A null page means the account holds none for that day.
  markRead(day, page) {
    const held = this.entries.get(day) ?? null;
    // A day this browser holds words for belongs to the claim, not to a read.
    if (held && held.needsPush && !held.read) return;
    // "Nothing here" arrives as a blank page at ZERO_STAMP, which loses every race, so a stamped entry
    // survives an empty read.
    const winner = winnerOf(page ?? blankPage(day), held?.page ?? null);
    this.entries.set(day, {
      page: winner,
      needsPush: Boolean(held?.needsPush) && winner === held.page,
      read: true,
    });
  }

  // A reply is a remote page arriving, not a debt cleared: newer typing may still be owed.
  markPushed(day, winner) {
    const held = this.entries.get(day) ?? null;
    // Strictly newer, not "at least as new": an acknowledged write comes back with the stamp we sent.
    const newerWaiting = Boolean(held?.needsPush) && compareStamps(held.page.stamp, winner.stamp) > 0;
    this.entries.set(day, {
      page: winnerOf(winner, held?.page ?? null),
      needsPush: newerWaiting,
      read: true,
    });
  }

  // Answers whether the device took the bytes.
  flush() {
    if (!this.storage) return false;
    try {
      // Union by day with what is on disk: a second tab holds its own copy and a blind write would drop
      // a day it never touched. Days both tabs hold are last-flush-wins.
      const merged = {};
      for (const [day, entry] of readEntries(this.storage, this.key)) merged[day] = entry;
      for (const [day, entry] of this.entries) merged[day] = entry;
      // The retention line applies to the union, so two tabs at different depths agree.
      const days = Object.keys(merged).sort();
      const oldestKept = days[Math.max(0, days.length - RETAIN_DAYS)] ?? '';
      const kept = {};
      for (const day of days) if (day >= oldestKept || merged[day].needsPush) kept[day] = merged[day];
      this.storage.setItem(this.key, JSON.stringify(kept));
      return true;
    } catch {
      return false;
    }
  }
}
